libs/checkpoint-postgres/src/index.ts

```diff
@@ -0,0 +1,330 @@
+import type { RunnableConfig } from "@langchain/core/runnables";
+import {
+  BaseCheckpointSaver,
+  type Checkpoint,
+  type CheckpointListOptions,
+  type CheckpointTuple,
+  type SerializerProtocol,
+  type PendingWrite,
+  type CheckpointMetadata,
+} from "@langchain/langgraph-checkpoint";
+import pg from "pg";
+
+const { Pool } = pg;
+
+interface CheckpointRow {
+  checkpoint: Buffer;
+  metadata: Buffer;
+  parent_checkpoint_id?: string;
+  thread_id: string;
+  checkpoint_id: string;
+  checkpoint_ns?: string;
+  type?: string;
+}
+
+interface WritesRow {
+  thread_id: string;
+  checkpoint_ns: string;
+  checkpoint_id: string;
+  task_id: string;
+  idx: number;
+  channel: string;
+  type?: string;
+  value?: Buffer;
+}
+
+export class PostgresSaver extends BaseCheckpointSaver {
+  private pool: pg.Pool;
+
+  protected isSetup: boolean;
+
+  constructor(pool: pg.Pool, serde?: SerializerProtocol) {
+    super(serde);
+    this.pool = pool;
+    this.isSetup = false;
+  }
+
+  static fromConnString(connString: string): PostgresSaver {
+    const pool = new Pool({ connectionString: connString });
+    return new PostgresSaver(pool);
+  }
+
+  protected async setup(): Promise<void> {
+    if (this.isSetup) {
+      return;
+    }
+
+    const client = await this.pool.connect();
+    try {
+      await client.query(`
+        CREATE TABLE IF NOT EXISTS checkpoints (
+          thread_id TEXT NOT NULL,
+          checkpoint_ns TEXT NOT NULL DEFAULT '',
+          checkpoint_id TEXT NOT NULL,
+          parent_checkpoint_id TEXT,
+          type TEXT,
+          checkpoint BYTEA NOT NULL,
+          metadata BYTEA NOT NULL,
+          PRIMARY KEY (thread_id, checkpoint_ns, checkpoint_id)
+        )`);
+      await client.query(`
+        CREATE TABLE IF NOT EXISTS writes (
```

Comment:
Perhaps a more specific table name like `checkpoint_writes`? Would also be nice if names were configurable like MongoDbSaver.

Reply:
Iirc, I just used the names as in sqlite checkpointer. But can sure give the tables a more telling name.